verifier-sdk/src/light_transaction.rs

```diff
@@ -190,19 +193,80 @@ impl<
 
     #[heap_neutral]
     #[inline(never)]
-    pub fn emit_indexer_transaction_event(&self) -> Result<()> {
+    pub fn emit_indexer_events(&self) -> Result<()> {
         #[cfg(all(target_os = "solana", feature = "mem-profiling"))]
         custom_heap::log_total_heap("pre assemble TransactionIndexerEvent");
 
         #[cfg(all(target_os = "solana", feature = "mem-profiling"))]
         custom_heap::log_total_heap("pre load MerkleTreeSet");
-        let merkle_tree_set = self.input.ctx.accounts.get_merkle_tree_set().load()?;
+        let merkle_tree_set_info = self.input.ctx.accounts.get_merkle_tree_set();
+        let merkle_tree_set = merkle_tree_set_info.load()?;
         #[cfg(all(target_os = "solana", feature = "mem-profiling"))]
         custom_heap::log_total_heap("post load MerkleTreeSet");
 
-        // Initialize the vector of leaves
-        let first_leaf_index =
-            state_merkle_tree_from_bytes(&merkle_tree_set.state_merkle_tree).next_index;
+        let state_merkle_tree = state_merkle_tree_from_bytes(&merkle_tree_set.state_merkle_tree);
+
+        // Emit the changelog event, twice: for the left and the right leaf.
+
+        let changelog_index_right = state_merkle_tree.current_changelog_index;
+        let changelog_index_left = changelog_index_right.saturating_sub(1);
+
+        let path_left = state_merkle_tree.changelog[changelog_index_left as usize].path;
+        let path_left = path_left
+            .iter()
+            .enumerate()
+            .map(|(level, node)| PathNode {
+                node: node.to_owned(),
+                index: (1 << (path_left.len() as u32 - level as u32))
+                    + (changelog_index_left as u32 >> level as u32),
+            })
+            .collect::<Vec<_>>();
+
+        let path_right = state_merkle_tree.changelog[changelog_index_right as usize].path;
+        let path_right = path_right
+            .iter()
+            .enumerate()
+            .map(|(level, node)| PathNode {
+                node: node.to_owned(),
+                index: (1 << (path_right.len() as u32 - level as u32))
```

Comment:
@ananas-block @sergeytimoshin @SwenSchaeferjohann Also, another thing which bothers me in this `index` code is that it produces quite weird results. Check out this playground output:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=33edd1055785bde1079bb644f0124730

```
Merkle path: 0
n: 420, index: 16
n: 69, index: 8
n: 56, index: 4
n: 11, index: 2
Merkle path: 1
n: 666, index: 17
n: 512, index: 8
n: 2137, index: 4
n: 30, index: 2
```

What the hell are even these indexes? Roots always marked with index 2? The only thing which looks right is that the leaves got incremented by 1 (from 16 to 17).

I look at the account compression code and I don't think I've made any mistakes:

https://github.com/solana-labs/solana-program-library/blob/6bd4832a0d66c63938c3ddb3e0a9af67ad6ec8cc/account-compression/programs/account-compression/src/events/changelog_event.rs#L49-L54

@SwenSchaeferjohann @sergeytimoshin I was struggling to find any example of a real `ChangeLogEventV1` from the SPL account compression programs, but do you have some by any chance? Can you notice any patterns of what the `index` value in changelogs is? Or point me how I can fetch some compressed NFT data to check it?

Reply:
Not sure, best check the indexer code for this. cc @sergeytimoshin

Reply:
it's being used in the indexer here https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/blob/5d69bfc6a8a7a2e4df8614a3ea5a010925a3e2d2/nft_ingester/src/program_transformers/bubblegum/db.rs#L16

Reply:
Yeah, I think I understand the purpose. I don't understand why root has index 2. 😅

Reply:
Good question haha